lib/pos_blockchain.go

```diff
@@ -0,0 +1,1134 @@
+		}
+		return nil
+	}
+	// Add to blockIndexByHash with status STORED only as we are not sure if it's valid yet.
+	_, err := bc.storeBlockInBlockIndex(block)
+	return errors.Wrap(err, "processBlockPoS: Problem adding block to block index: ")
+}
+
+// storeValidateFailedBlockWithWrappedError is a helper function that takes in a block and an error and
+// stores the block in the block index with status VALIDATE_FAILED. It returns the resulting BlockNode.
+func (bc *Blockchain) storeValidateFailedBlockWithWrappedError(block *MsgDeSoBlock, outerErr error) (*BlockNode, error) {
+	blockNode, innerErr := bc.storeValidateFailedBlockInBlockIndex(block)
+	if innerErr != nil {
+		return nil, errors.Wrapf(innerErr, "storeValidateFailedBlockWithWrappedError: Problem adding validate failed block to block index: %v", outerErr)
+	}
+	return blockNode, nil
+}
+
+// validateAndIndexBlockPoS performs all validation checks for a given block and adds it to the block index with
+// the appropriate status.
+//  1. If the block is already VALIDATE_FAILED, we return the BlockNode as-is without perform further validations and
+//     throw an error.
+//  2. If the block is already VALIDATED, we return the BlockNode as-is without performing further validations and no error.
+//  3. We check if its parent is VALIDATE_FAILED, if so we add the block to the block index with status VALIDATE_FAILED
+//     and throw an error.
+//  4. If its parent is NOT VALIDATED and NOT VALIDATE_FAILED, we recursively call this function on its parent.
+//  5. If after calling this function on its parent, the parent is VALIDATE_FAILED, we add the block to the block index
+//     with status VALIDATE_FAILED and throw an error.
+//  6. If after calling this function on its parent, the parent is VALIDATED, we perform all other validations on the block.
+//
+// The recursive function's invariant is described as follows:
+//   - Base case: If block is VALIDATED or VALIDATE_FAILED, return the BlockNode as-is.
+//   - Recursive case: If the block is not VALIDATED or VALIDATE_FAILED in the blockIndexByHash, we will perform all
+//     validations and add the block to the block index with the appropriate status (VALIDATED OR VALIDATE_FAILED) and
+//     return the new BlockNode.
+//   - Error case: Something goes wrong that doesn't result in the block being marked VALIDATE or VALIDATE_FAILED. In
+//     this case, we will add the block to the block index with status STORED and return the BlockNode.
+func (bc *Blockchain) validateAndIndexBlockPoS(block *MsgDeSoBlock) (*BlockNode, error) {
+	blockHash, err := block.Header.Hash()
+	if err != nil {
+		return nil, errors.Wrapf(err, "validateAndIndexBlockPoS: Problem hashing block %v", block)
+	}
+
+	// Base case - Check if the block is validated or validate failed. If so, we can return early.
+	blockNode, exists := bc.blockIndexByHash[*blockHash]
+	if exists && (blockNode.IsValidateFailed() || blockNode.IsValidated()) {
+		return blockNode, nil
```

Comment:
Hmm don't you want to return an error if the block is invalid?

Reply:
No - This function only returns an error if we hit an unexpected error such as an error reading from badger or something. The caller is returned the block node and can decide what to do from there. If the caller receives an error, it means that something unexpected happened and we don't know if the block is truly invalid or just had some weird intermittent error.